### PR TITLE
implement the base library for compacting logs

### LINKED ISSUE #17631: compact_log_backup: implement the basic library for compacting
## Development Task

Add a subcrate that can be used to compact logs in `components`.

### PR DESCRIPTION
**Pull request by YuJuncen**
### What is changed and how it works?

Issue Number: Close #17631

What's Changed:

```commit-message
Added a new crate named `compact-log-backup`. Now it can merge some log files generated by log backup and make them become SSTs.
```

The directory hierarchy:

```
./components/compact-log-backup/
├── Cargo.toml
└── src
    ├── compaction ← Things about compacting logs, including metadata and the logic of executing an compaction.
    │   ├── collector.rs ← Give some files, what compaction we will make?
    │   ├── exec.rs ← How we read and sort log entries and make them SSTs?
    │   ├── meta.rs ← How to describe a compaction?
    │   └── mod.rs
    ├── errors.rs ← Error definitions.
    ├── exec_hooks ← Common hooks.
    │   ├── checkpoint.rs ← Skip compactions that already done.
    │   ├── consistency.rs ← Lock the storage to keep it consistent.
    │   ├── mod.rs
    │   ├── observability.rs ← Expose metrics and print logs.
    │   └── save_meta.rs ← Save the compaction metadata to the external storage.
    ├── execute
    │   ├── hooking.rs ← What hooks we have?
    │   ├── mod.rs ← The controller of a compaction over time.
    │   └── test.rs
    ├── lib.rs
    ├── source.rs ← Data source, for now only logs are supported.
    ├── statistic.rs 
    ├── storage.rs ← Things helps to read the backup storage.
    ├── test_util.rs
    └── util.rs
```

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Leavrth — APPROVED** [review]

File: components/compact-log-backup/src/execute/test.rs [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        checkpoint::Checkpoint, consistency::StorageConsistencyGuard, save_meta::SaveMeta,
+    },
+    execute::hooking::{CId, ExecHooks, SubcompactionFinishCtx},
+    storage::LOCK_PREFIX,
+    test_util::{gen_step, CompactInMem, KvGen, LogFileBuilder, TmpStorage},
+    ErrorKind,
+};
+
+#[derive(Clone)]
+struct CompactionSpy(Sender<SubcompactionResult>);
```
[the remark this answers is not in the document]
Comment (YuJuncen) [reply components/compact-log-backup/src/execute/test.rs:32]: I guess the file itself is in the mod `test`. Do you meaning the `test_utli` mod?

File: components/compact-log-backup/src/compaction/collector.rs [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+
+        self.stat.bytes_in += file.file_real_size;
+        self.stat.files_in += 1;
+
+        match self.items.entry(key) {
+            Entry::Occupied(mut o) => {
+                let key = *o.key();
+                let u = o.get_mut();
+                u.add_file(file);
+                if u.size > self.cfg.subcompaction_size_threshold {
```
[the remark this answers is not in the document]
Comment (YuJuncen) [reply components/compact-log-backup/src/compaction/collector.rs:98]: Perhaps. We may add something like memory quota in the future. But for now in fact we cannot do better.

File: components/compact-log-backup/src/exec_hooks/checkpoint.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        self.load(
+            cx.storage,
+            &format!("{}/{}", cx.this.out_prefix, META_OUT_REL),
+        )
+        .await
+    }
+
+    fn before_a_subcompaction_start(&mut self, _cid: CId, cx: SubcompactionStartCtx<'_>) {
+        let hash = cx.subc.crc64();
+        if self.loaded.contains(&hash) {
```
[the remark this answers is not in the document]
Comment (YuJuncen) [reply components/compact-log-backup/src/exec_hooks/checkpoint.rs:75]: Then the second one cannot be executed. Thankfully its input will probably be saved as the final subcompaction won't be written.

File: components/compact-log-backup/src/util.rs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    pub fn new(yield_every: usize) -> Self {
+        Self {
+            work_count: 0,
+            yield_every,
+        }
+    }
+
+    /// Finishing one tiny task. This will yield the current carrier thread
+    /// when needed.
+    pub fn step(&mut self) -> Step {
```
Comment (3pointer) [inline components/compact-log-backup/src/util.rs:50]: Can we use a wrapper of `tokio::task::yield_now().await;` here?

File: components/compact-log-backup/src/compaction/exec.rs [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+        let begin = Instant::now();
+        let (sorted_items, cdiff) = self.process_input(items).await;
+        self.compact_stat.sort_duration += begin.saturating_elapsed();
+        if let Some(v) = result.expected_crc64.as_mut() {
+            *v ^= cdiff.crc64xor_diff;
+        }
+        result.expected_keys -= cdiff.removed_key;
+        result.expected_size -= cdiff.decreaed_size;
+
+        if sorted_items.is_empty() {
```
Comment (3pointer) [inline components/compact-log-backup/src/compaction/exec.rs:353]: move this into L347

**Review by 3pointer — APPROVED** [review]
rest LGTM

**Review by iosmanthus — APPROVED** [review]

**Pull request merged**